posts/opaleyes-api-breakage-policy.md

```diff
@@ -0,0 +1,385 @@
+often left unaccounted for: it takes effort even to discover what
+changes need to be made!
+
+In my experience of being on the receiving end of API breakage the
+latter cost is actually the more severe.  Once I know what change
+needs to be made I can come up with a plan and accurately assess the
+effort involved.  On the other hand, if the only information that I
+receive is a compile error, and no breadcrumbs pointing me in the
+direction of a solution, then I am left staring at a long-tailed
+distribution: for all I know fixing the breakage could take me a
+minute, an hour or a day.  Under such uncertainty I am disinclined to
+even start.
+
+### Why break?
+
+The aim of Opaleye is to be of service to users, so if API breakage
+imposes a cost on users then why break the API at all?  Firstly, as
+sole maintainer I need to keep my burden light, and striving for a
+small API surface area is one way of doing that.  Secondly, our
+understanding of Opaleye best practice evolves over time, and I want
+to nudge users towards best practice.  Thirdly, I want the API to be
+approachable, and removing old, deprecated parts of the API helps that
+end.
+
+This rationale is the justification for occasional API breakage.  The
+alternative is to keep old, discouraged parts of the API around, but
+to warn against them, deprioritise them in documentation and avoid
+mentioning them in polite conversation.  That alternative is a valid
+approach too, perhaps even a desirable approach if one has the
+resources to deal with the mental overhead, but I don't feel that I
+do.
+
+So, the aim of the Opaleye breakage policy is to strike a balance
+between keeping the API small, simple, easy to understand and
+maintainable on the one hand, and reducing the costs of churn on the
+user on the other.  Later versions will inevitably become incompatible
+with user code written against older versions, but there is much that
+can be done to mitigate the costs on the user, to "hold the user's
+hand" through the process, and make their life easier, some of which
+the rest of this article will describe.
+
+## Avoid flag day
+
+The single biggest breakage cost that an API can impose is the need
+for a "[flag
+day](https://en.wikipedia.org/wiki/Flag_day_(computing))", that is "a
+change which requires a complete conversion of a sizable body of
+software".  The cost of a flag day is not just borne by the users of
+the API individually; the changes that lead to flag days make software
+ecosystems rigid and inflexible and impose coordination costs across
+*the entire ecosystem*!  Flag days force the entire ecosystem to
+change in lock step.
+
+I work very hard to avoid "flag days" caused by Opaleye API changes.
+Specifically, if a user has written code against Opaleye major version
+`N`[^two-components] that is broken by Opaleye `N+1`, it should be
+possible to write different code *with the same functionality* that
+works against versions `N` and `N+1` at the same time.  For example,
```

Comment:
Perhaps it would be ideal if a restriction of dependency bounds and changes to the library itself would never be done in the same release?

Reply:
This is a really interesting question! Yes, without a compatibility module it's rather difficult to support aeson 1.5 and 2.0 simultaneously. Your idea of

> restriction of dependency bounds and changes to the library itself would never be done in the same release

seems worth considering further. On the one hand, I don't think it ever makes sense to *reduce* the dependency range, but it might make sense to shift it up. For example, if Opaleye `N` was compatible with `postgresql-simple` `0.4` and `0.5`, but not `0.6` on its release, but it was not possible for Opaleye to support all three of those versions at the same time, then perhaps Opaleye `N+1` could support `0.5` and `0.6` without making any other breaking changes.

On the other hand, if `postgresql-simple` had the same breakage policy as Opaleye then perhaps it would make sense to update Opaleye `N` to `N+1` including breaking changes that satisfy the fundamental principle, along with changing the `postgresql-simple` dependency to `0.5` or `0.6`.  In that case users could write code that targets Opaleye `N` and `postgresql-simple` `0.5` *and* Opaleye `N+1` and `postgresql-simple` `0.6` at the same time, and the fundamental property is preserved by the composition. (The fundamental property has a "compositional" property.)

There's probably a lot more that could be said here!  I'll have a think about it. It could be very useful subject matter for a follow-on article.